File: Lab_Assignments/filesystems/flock.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/file.h>
#include <sys/stat.h>
#include <unistd.h>
#include <fcntl.h>

int main() {
    char filename[256];
    int fd;
    struct stat st;

    printf("Enter file name: ");
    scanf("%s", filename);

    fd = open(filename, O_RDWR);
    if (fd < 0) {
        perror("open");
        exit(1);
    }

    if (fstat(fd, &st) < 0) {
        perror("fstat");
        close(fd);
        exit(1);
    }

    // Mandatory lock check: regular file + set-GID + group execute off
    if (S_ISREG(st.st_mode) && 
        (st.st_mode & S_ISGID) && 
        !(st.st_mode & S_IXGRP)) {
        printf("Mandatory locking is enabled on this file.\n");
    } else {
        printf("Only advisory locking is possible.\n");
    }

    printf("Trying to acquire exclusive lock...\n");
    if (flock(fd, LOCK_EX) < 0) {
        perror("flock");
        close(fd);
        exit(1);
    }

    printf("Lock acquired. Press Enter to release.\n");
    getchar(); // consume leftover newline
    getchar(); // wait for user

    flock(fd, LOCK_UN);
    printf("Lock released.\n");

    close(fd);
    return 0;
}

File: PA/last_quest/flock.py
import fcntl
import time
import os

filename = "demo_flock.txt"

with open(filename, "w+") as f:
    print(f"Process {os.getpid()} waiting for flock...")
    fcntl.flock(f, fcntl.LOCK_EX)  # Acquire exclusive lock
    print(f"Process {os.getpid()} got the flock lock!")
    f.write(f"Locked by {os.getpid()}\n")
    time.sleep(5)
    print("Releasing flock lock...")
    fcntl.flock(f, fcntl.LOCK_UN)
